src/hotspot/share/runtime/arguments.cpp

```diff
@@ -3768,6 +3768,9 @@ void Arguments::set_compact_headers_flags() {
   if (UseCompactObjectHeaders && FLAG_IS_DEFAULT(hashCode)) {
     hashCode = 6;
   }
+  if (UseCompactObjectHeaders && FLAG_IS_DEFAULT(CompressedClassSpaceSize)) {
+    FLAG_SET_DEFAULT(CompressedClassSpaceSize, 512 * M);
+  }
```

Comment:
Still a bit ugly, since again hardcoded.

What I would do is:

1) move the call to `CompressedKlassPointers::pre_initialize()` at line 3789:

```
  if (UseCompressedClassPointers) {
    CompressedKlassPointers::pre_initialize();
  }
```

down into ``  from apply_ergo (just below) to before you set the CCSS.

Then, set the CCS with FLAG_SET_DEFAULT(CompressedClassSpaceSize, 512 * M);